src/Compiler/Checking/TypeRelations.fs

```diff
@@ -134,50 +134,57 @@ let rec TypeFeasiblySubsumesType ndeep g amap m ty1 canCoerce ty2 =
              let join m x = 
-                 if TypeFeasiblySubsumesType 0 g amap m x CanCoerce maxTy then maxTy
-                 elif TypeFeasiblySubsumesType 0 g amap m maxTy CanCoerce x then x
-                 else errorR(Error(FSComp.SR.typrelCannotResolveImplicitGenericInstantiation((DebugPrint.showType x), (DebugPrint.showType maxTy)), m)); maxTy
+                 if TypeFeasiblySubsumesType 0 g amap m x CanCoerce maxTy then maxTy, isRefined
+                 elif TypeFeasiblySubsumesType 0 g amap m maxTy CanCoerce x then x, true
+                 else errorR(Error(FSComp.SR.typrelCannotResolveImplicitGenericInstantiation((DebugPrint.showType x), (DebugPrint.showType maxTy)), m)); maxTy, isRefined
              // Don't continue if an error occurred and we set the value eagerly 
-             if tp.IsSolved then maxTy, m else
+             if tp.IsSolved then (maxTy, isRefined), m else
              match tpc with 
              | TyparConstraint.CoercesTo(x, m) -> 
                  join m x, m
              | TyparConstraint.MayResolveMember(_traitInfo, m) -> 
-                 maxTy, m
+                 (maxTy, isRefined), m
              | TyparConstraint.SimpleChoice(_, m) -> 
                  errorR(Error(FSComp.SR.typrelCannotResolveAmbiguityInPrintf(), m))
-                 maxTy, m
+                 (maxTy, isRefined), m
              | TyparConstraint.SupportsNull m -> 
-                 maxTy, m
+                 (maxTy, isRefined), m
              | TyparConstraint.SupportsComparison m -> 
                  join m g.mk_IComparable_ty, m
              | TyparConstraint.SupportsEquality m -> 
-                 maxTy, m
+                 (maxTy, isRefined), m
              | TyparConstraint.IsEnum(_, m) -> 
                  errorR(Error(FSComp.SR.typrelCannotResolveAmbiguityInEnum(), m))
-                 maxTy, m
+                 (maxTy, isRefined), m
              | TyparConstraint.IsDelegate(_, _, m) -> 
                  errorR(Error(FSComp.SR.typrelCannotResolveAmbiguityInDelegate(), m))
-                 maxTy, m
+                 (maxTy, isRefined), m
              | TyparConstraint.IsNonNullableStruct m -> 
                  join m g.int_ty, m
              | TyparConstraint.IsUnmanaged m ->
                  errorR(Error(FSComp.SR.typrelCannotResolveAmbiguityInUnmanaged(), m))
-                 maxTy, m
+                 (maxTy, isRefined), m
              | TyparConstraint.RequiresDefaultConstructor m -> 
-                 maxTy, m
+                 (maxTy, isRefined), m
              | TyparConstraint.IsReferenceType m -> 
-                 maxTy, m
+                 (maxTy, isRefined), m
              | TyparConstraint.DefaultsTo(_priority, _ty, m) -> 
-                 maxTy, m)
+                 (maxTy, isRefined), m)
+
+    match tp.Kind with
+    | TyparKind.Type ->
+        if not isRefined then
+            warning(Error(FSComp.SR.typrelNeverRefinedAwayFromTop(), m))
```

Comment:
Can you make this an "Information" diagnostics for the current PR ?
And increase to warning later && behind a language version feature.

Reply:
Yeah, +1 to what Tomas said - since it's a new diagnostic, it needs to be under language flag (i.e. only start showing up for next language version).

Reply:
Even though it's off by default and you have to take specific action to enable it, and that action is the same regardless of your current language version?

Reply:
Yes, this is a .NET SDK-wide guideline we try to adhere to. You tie new features to language version so that you have a convenient hook to gradually increase the diagnostic level, and so that users have a predictable, common place to expect changes - the learned behavior is "when I bump my TFM I should expect new diagnostics", and as much as possible no other event should cause new diagnostics.

This is primarily to help ensure that users can float to the latest SDKs as much as possible without causing new errors (within reason).

Reply:
Fair enough, although I don't see how the reasoning applies here - floating to a new SDK will of course not cause this diagnostic to appear unless you have explicitly already requested that it appear. I'll try and work out how to do that when I'm at my computer at some point.

Reply:
Most of the times we want to have most of the new logic to be behind language version check (even if it doesn't bring any visible changes), so compilation pipeline stays the same in each langauge version, otherwise it defeats the purpose of having language versions in general.

Reply:
Have I misunderstood something about language versions - the compilation pipeline will change as you move SDKs regardless, won't it? The change would merely be suppressing one output of the compilation.

Reply:
Does that mean there's a feature for "introduced the diagnostic at Information level", and then a subsequent feature for "increased the diagnostic to Warning level"?

Reply:
> Does that mean there's a feature for "introduced the diagnostic at Information level", and then a subsequent feature for "increased the diagnostic to Warning level"?

Likely yes, we don't really have more granular mechanism for it. Would've been nice to have it though.

Reply:
Raised https://github.com/dotnet/fsharp/pull/15315 to reserve the feature flag.